src/envs/behavior.py

```diff
@@ -244,60 +312,92 @@ def _classifier(s: State, o: Sequence[Object]) -> bool:
                 return bddl_ground_atom.get_value()
             if arity == 2:
                 assert len(o) == 2
-                ig_obj = self._object_to_ig_object(o[0])
-                other_ig_obj = self._object_to_ig_object(o[1])
+                ig_obj = self.object_to_ig_object(o[0])
+                other_ig_obj = self.object_to_ig_object(o[1])
                 bddl_partial_ground_atom = bddl_predicate.STATE_CLASS(ig_obj)
-                bddl_partial_ground_atom.initialize(self._env.simulator)
+                bddl_partial_ground_atom.initialize(
+                    self.behavior_env.simulator)
                 return bddl_partial_ground_atom.get_value(other_ig_obj)
+
             raise ValueError("BDDL predicate has unexpected arity.")
 
         return _classifier
 
+    def _graspable_classifier(self, state: State,\
+        objs: Sequence[Object]) -> bool:
+        # Check allclose() here for uniformity with
+        # _create_classifier_from_bddl
+        assert state.allclose(self._current_ig_state_to_state())
+        assert len(objs) == 1
+        ig_obj = self.object_to_ig_object(objs[0])
+        lo, hi = ig_obj.states[object_states.AABB].get_value()
+        volume = get_aabb_volume(lo, hi)
+        return volume < 0.3 * 0.3 * 0.3 and not ig_obj.main_body_is_fixed
+
+
+    def _reachable_classifier(self, state: State,\
+        objs: Sequence[Object]) -> bool:
+        # Check allclose() here for uniformity with
+        # _create_classifier_from_bddl
+        assert state.allclose(self._current_ig_state_to_state())
+        assert len(objs) == 2
+        ig_obj = self.object_to_ig_object(objs[0])
+        ig_other_obj = self.object_to_ig_object(objs[1])
+        return (np.linalg.norm(  # type: ignore
+            np.array(ig_obj.get_position()) -
+            np.array(ig_other_obj.get_position())) < 2)
+
+    def _reachable_nothing_classifier(self, state: State,\
+        objs: Sequence[Object]) -> bool:
+        # Check allclose() here for uniformity with _create_classifier_from_bddl
+        assert state.allclose(self._current_ig_state_to_state())
+        assert len(objs) == 1
+        for obj in state:
+            if self._reachable_classifier(
+                    state=state, objs=[obj, objs[0]]) and (obj != objs[0]):
+                return False
+        return True
+
     def _get_grasped_objects(self, state: State) -> Set[Object]:
         grasped_objs = set()
         for obj in state:
-            ig_obj = self._object_to_ig_object(obj)
-            if any(self._env.robots[0].is_grasping(ig_obj)):
+            ig_obj = self.object_to_ig_object(obj)
+
+            # NOTE: The below block is necessary because somehow the body_id
+            # is sometimes a 1-element list...
+            if isinstance(ig_obj.body_id, list):
+                assert len(ig_obj.body_id) == 1
+                ig_obj.body_id = ig_obj.body_id[0]
+
+            if any(self.behavior_env.robots[0].is_grasping(ig_obj.body_id)):
```

Comment:
just wondering, why is the `any` necessary? what's the return of `is_grasping`?

Reply:
it returns an `np.ndarray` by design it seems

Reply:
let's use `np.any` then?